src/bin/miri.rs

```diff
@@ -78,14 +78,22 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
 
             let early_dcx = EarlyDiagCtxt::new(tcx.sess.opts.error_format);
             init_late_loggers(&early_dcx, tcx);
-            if !tcx.crate_types().contains(&CrateType::Executable) {
-                tcx.dcx().fatal("miri only makes sense on bin crates");
-            }
 
             let (entry_def_id, entry_type) = if let Some(entry_def) = tcx.entry_fn(()) {
                 entry_def
             } else {
-                tcx.dcx().fatal("miri can only run programs that have a main function");
+                let sym = tcx.exported_symbols(LOCAL_CRATE).iter().find_map(|(sym, _)| {
+                    if sym.symbol_name_for_local_instance(tcx).name == "miri_start" {
+                        Some(sym)
+                    } else {
+                        None
+                    }
+                });
+                if let Some(ExportedSymbol::NonGeneric(id)) = sym {
+                    (*id, EntryFnType::Start)
+                } else {
+                    tcx.dcx().fatal("miri can only run programs that have a main function or one exported with symbol miri_start");
```

Comment:
the error message should probably differentiate between no symbol found and a generic symbol found

Reply:
I'd then reword the not found one to
> Miri can only run programs that have a main function or export the symbol `miri_start` (for example, with `#[no_mangle] fn miri_start()`)"